README.md
Librato Metrics
=======

[![Gem Version](https://badge.fury.io/rb/librato-metrics.png)](http://badge.fury.io/rb/librato-metrics) [![Build Status](https://secure.travis-ci.org/librato/librato-metrics.png?branch=master)](http://travis-ci.org/librato/librato-metrics) [![Code Climate](https://codeclimate.com/github/librato/librato-metrics.png)](https://codeclimate.com/github/librato/librato-metrics)

A convenient Ruby wrapper for the Librato Metrics API.

---

## Important note on breaking change

**NOTE:** Starting with version 2.1.0 librato-metrics requires a Librato account that [supports tagged metrics](https://www.librato.com/docs/kb/faq/account_questions/tags_or_sources/). If your Librato account doesn't yet support tagged metrics please use the [1.6.1 version](https://rubygems.org/gems/librato-metrics/versions/1.6.1).

---

This gem provides granular control for scripting interactions with the Metrics core API. It is well suited for integrations, scripts, workers & background jobs. If you want to submit from a web app, take at look at [librato-rails](https://github.com/librato/librato-rails) and/or [librato-rack](https://github.com/librato/librato-rack).

## Installation

In your shell:

    gem install librato-metrics

Then, in your application or script:

    require 'librato/metrics'

### Optional steps

For best performance we recommend installing [yajl-ruby](https://github.com/brianmario/yajl-ruby):

    gem install yajl-ruby

If you are using jruby, you need to ensure [jruby-openssl](https://github.com/jruby/jruby-ossl) is available:

    gem install jruby-openssl

## Quick Start

If you are looking for the quickest possible route to getting a data into Metrics, you only need two lines:

    Librato::Metrics.authenticate 'email', 'api_key'
    Librato::Metrics.submit my_metric: { value: 42, tags: { host: 'localhost' } }

While this is all you need to get started, if you are sending a number of metrics regularly a queue may be easier/more performant so read on...


## Authentication

Make sure you have [an account for Librato](https://metrics.librato.com/) and then authenticate with your email and API key (on your account page):

    Librato::Metrics.authenticate 'email', 'api_key'


## Sending Measurements

A measurement includes a metric name, value, and one or more tags. Tags include a name/value pair that describe a particular data stream. Each unique tag set creates an individual metric stream which can later be filtered and aggregated along.

Queue up a simple metric named `temperature`:

    queue = Librato::Metrics::Queue.new
    queue.add temperature: {value: 77, tags: { city: 'oakland' }}
    queue.submit

### Top-Level Tags

You can initialize `Queue` and/or `Aggregator` with top-level tags that will be applied to every measurement:

    queue = Librato::Metrics::Queue.new(tags: { service: 'auth', environment: 'prod', host: 'auth-prod-1' })
    queue.add my_metric: 10
    queue.submit

### Per-Measurement Tags

Optionally, you can submit per-measurement tags by passing a tags Hash when adding measurements:

    queue.add my_other_metric: { value: 25, tags: { db: 'rr1' } }
    queue.submit

For more information, visit the [API documentation](https://www.librato.com/docs/api/#create-a-measurement).


## Querying Metrics

Get name and properties for all metrics you have in the system:

    metrics = Librato::Metrics.metrics

Get only metrics whose name includes `time`:

    metrics = Librato::Metrics.metrics name: 'time'


## Retrieving Measurements

Get the series for `exceptions` in **production** grouped by **sum** within the **last hour**:

```ruby
query = {
  resolution: 1,
  duration: 3600,
  group_by: "environment",
  group_by_function: "sum",
  tags_search: "environment=prod*"
}
Librato::Metrics.get_series :exceptions, query
```

For more information, visit the [API documentation](https://www.librato.com/docs/api/#retrieve-a-measurement).


## Aggregate Measurements

If you are measuring something very frequently e.g. per-request in a web application (order mS)  you may not want to send each individual measurement, but rather periodically send a [single aggregate measurement](https://www.librato.com/docs/api/#gauge-specific-parameters), spanning multiple seconds or even minutes. Use an `Aggregator` for this.

Aggregate a simple gauge metric named `response_latency`:

    aggregator = Librato::Metrics::Aggregator.new
    aggregator.add response_latency: 85.0
    aggregator.add response_latency: 100.5
    aggregator.add response_latency: 150.2
    aggregator.add response_latency: 90.1
    aggregator.add response_latency: 92.0

Which would result in a gauge measurement like:

    {name: "response_latency", count: 5, sum: 517.8, min: 85.0, max: 150.2}

You can specify a source during aggregate construction:

    aggregator = Librato::Metrics::Aggregator.new(tags: { service: 'auth', environment: 'prod', host: 'auth-prod-1' })

You can aggregate multiple metrics at once:

    aggregator.add app_latency: 35.2, db_latency: 120.7

Send the currently aggregated metrics to Metrics:

    aggregator.submit

## Benchmarking

If you have operations in your application you want to record execution time for, both `Queue` and `Aggregator` support the `#time` method:

    aggregator.time :my_measurement do
      # do work...
    end

The difference between the two is that `Queue` submits each timing measurement individually, while `Aggregator` submits a single timing measurement spanning all executions.

If you need extra attributes for a `Queue` timing measurement, simply add them on:

    queue.time :my_measurement do
      # do work...
    end

## Annotations

Annotation streams are a great way to track events like deploys, backups or anything else that might affect your system. They can be overlaid on any other metric stream so you can easily see the impact of changes.

At a minimum each annotation needs to be assigned to a stream and to have a title. Let's add an annotation for deploying `v45` of our app to the `deployments` stream:

    Librato::Metrics.annotate :deployments, 'deployed v45'

There are a number of optional fields which can make annotations even more powerful:

    Librato::Metrics.annotate :deployments, 'deployed v46', source: 'frontend',
        start_time: 1354662596, end_time: 1354662608,
        description: 'Deployed 6f3bc6e67682: fix lotsa bugs…'

You can also automatically annotate the start and end time of an action by using `annotate`'s block form:

    Librato::Metrics.annotate :deployments, 'deployed v46' do
      # do work..
    end

More fine-grained control of annotations is available via the `Annotator` object:

    annotator = Librato::Metrics::Annotator.new

    # list annotation streams
    streams = annotator.list

    # fetch a list of events in the last hour from a stream
    annotator.fetch :deployments, start_time: (Time.now.to_i-3600)

    # delete an event
    annotator.delete_event 'deployments', 23

See the documentation of `Annotator` for more details and examples of use.

## Auto-Submitting Metrics

Both `Queue` and `Aggregator` support automatically submitting measurements on a given time interval:

    # submit once per minute
    timed_queue = Librato::Metrics::Queue.new(autosubmit_interval: 60)

    # submit every 5 minutes
    timed_aggregator = Librato::Metrics::Aggregator.new(autosubmit_interval: 300)

`Queue` also supports auto-submission based on measurement volume:

    # submit when the 400th measurement is queued
    volume_queue = Librato::Metrics::Queue.new(autosubmit_count: 400)

These options can also be combined for more flexible behavior.

Both options are driven by the addition of measurements. *If you are adding measurements irregularly (less than once per second), time-based submission may lag past your specified interval until the next measurement is added.*

If your goal is to collect metrics every _x_ seconds and submit them, [check out this code example](https://github.com/librato/librato-metrics/blob/master/examples/submit_every.rb).

## Setting Metric Properties

Setting custom [properties](https://www.librato.com/docs/api/#metric-attributes) on your metrics is easy:

    # assign a period and default color
    Librato::Metrics.update_metric :temperature, period: 15, attributes: { color: 'F00' }

## Deleting Metrics

If you ever need to remove a metric and all of its measurements, doing so is easy:

    # delete the metrics 'temperature' and 'humidity'
    Librato::Metrics.delete_metrics :temperature, :humidity

You can also delete using wildcards:

    # delete metrics that start with cpu. except for cpu.free
    Librato::Metrics.delete_metrics names: 'cpu.*', exclude: ['cpu.free']

Note that deleted metrics and their measurements are unrecoverable, so use with care.

## Using Multiple Accounts Simultaneously

If you need to use metrics with multiple sets of authentication credentials simultaneously, you can do it with `Client`:

    joe = Librato::Metrics::Client.new
    joe.authenticate 'email1', 'api_key1'

    mike = Librato::Metrics::Client.new
    mike.authenticate 'email2', 'api_key2'

All of the same operations you can call directly from `Librato::Metrics` are available per-client:

    # list Joe's metrics
    joe.metrics

There are two ways to associate a new queue with a client:

    # these are functionally equivalent
    joe_queue = Librato::Metrics::Queue.new(client: joe)
    joe_queue = joe.new_queue

Once the queue is associated you can use it normally:

    joe_queue.add temperature: { value: 65.2, tags: { city: 'san francisco' } }
    joe_queue.submit

## Thread Safety

The `librato-metrics` gem currently does not do internal locking for thread safety. When used in multi-threaded applications, please add your own [mutexes](http://www.ruby-doc.org/core-2.0/Mutex.html) for sensitive operations.

## More Information

`librato-metrics` is sufficiently complex that not everything can be documented in the README. Additional options are documented regularly in the codebase. You are encouraged to take a quick look through the [source](https://github.com/librato/librato-metrics) for more.

We also maintain a set of [examples of common uses](https://github.com/librato/librato-metrics/tree/master/examples) and appreciate contributions if you have them.

## Contribution

* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet.
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it.
* Fork the project and submit a pull request from a feature or bugfix branch.
* Please review our [code conventions](https://github.com/librato/librato-metrics/wiki/Code-Conventions).
* Please include specs. This is important so we don't break your changes unintentionally in a future version.
* Please don't modify the gemspec, Rakefile, version, or changelog. If you do change these files, please isolate a separate commit so we can cherry-pick around it.

## Copyright

Copyright (c) 2011-2017 [Solarwinds, Inc.](http://www.solarwinds.com) See LICENSE for details.


CHANGELOG.md
## Changelog

### Version 1.6.2
* Relax Faraday version constraint (#143) (Jason Rudolph)

### Version 1.6.1
* Fix bugs with listing sources (#116)

### Version 1.6.0
* Add HTTP proxy support (#112) (Genki Sugawara)

### Version 1.5.1
* Fix bug causing incompatible dates when annotating with block form (#109)

### Version 1.5.0
* Add #get_composite for easier fetching of composite measurements

### Version 1.4.0
* Add support for snapshots

### Version 1.3.2
* Fix queue autosubmission to fire if needed after #merge! calls

### Version 1.3.1
* Fix auto-chunking for large measurements sets with a global source

### Version 1.3.0
* Add support for working with sources as a first-class entity

### Version 1.2.0
* Give metric-facing methods more explicit names & deprecate priors
* Documentation improvements

### Version 1.1.1
* Move gem sign code to rake task, fixes bug bundling in some environments

### Version 1.1.0
* Add ability to update annotation events
* Add ability to fetch annotation events
* Add block form of annotation
* Add metric batch update support
* Add support for pattern-based metric deletes
* Add set of code examples
* Sign gem when building
* Documentation improvements

### Version 1.0.4
* Ensure sane default timeouts for all requests

### Version 1.0.3
* Fix bug where retries of POST requests could 400
* Network related exceptions capture response state better

### Version 1.0.2
* Fix bug with some versions of MultiJson (Thomas Dippel)
* Use delegation for JSON handling
* Improve integration tests

### Version 1.0.1
* Fix Forwardable dependency loading bug

### Version 1.0.0
* Add support for annotation submission, listing, management
* Auto-convert Time objects anywhere a time is accepted
* Don't raise exception anymore for empty queue submission

### Version 0.7.5
* Catch a broader range of connection failures for retrying
* Add Metrics.faraday_adapter config option (Mathieu Ravaux)

### Version 0.7.4
* Support global measure_time option for Queues/Aggregators
* Support all versions of multi_json so we can relax version constraint

### Version 0.7.3
* Allow prefixes to be changed after instantiation on Queues/Aggregators

### Version 0.7.2
* Extend prefix option support to Aggregators

### Version 0.7.1
* Add prefix option to Queues

### Version 0.7.0
* Add ability to update metric properties (Christoph Bünte)
* Add ability to merge queue and aggregator data into a queue
* Aggregator supports custom source by measurement
* Add option to clear queued measurements after failed submit
* Custom user agent support
* Documentation improvements

### Version 0.6.1
* Loosen restrictions to older versions of faraday and multi_json
* Fix symbol casting issue in jruby with metric delete
* client#new_queue now respects passed options
* Queue objects support default source properly

### Version 0.6.0
* Add Aggregator class for aggregating measurements client-side
* Queue and Aggregator can auto-submit on a time interval
* Queue can auto-submit on a specified volume of measurements
* Support deleting individual metrics
* Validate user-specified measurement times
* Update to MultiJSON 1.3 syntax
* Run tests for rubinius and jruby in both 1.8 and 1.9 modes
* Include request body in output for failed requests
* Documentation improvements

### Version 0.5.0
* Support using multiple accounts simultaneously via Client
* Switch network library to faraday for broader platform support and flexibility
* Automatically break large submissions into multiple requests for better performance
* Automatic retry support
* Consolidate connection functions in Connection
* Documentation improvements

### Version 0.4.3
* Bump excon to 0.13x to fix proxy support

### Version 0.4.2
* Fix SSL verify peer issues with JRuby (Sean Porter)

### Version 0.4.1
* Fix issues with auth encoding whitespace

### Version 0.4.0
* Add ability to set agent_identifier for use with developer program (Sean Porter)
* Documentation improvements

### Version 0.3.1
* Upgrade excon to 0.9.5 to fix intermittent socket errors

### Version 0.3.0
* Add auto-pagination support to metric listing (Nuno Valente)
* Add #size/#length to Queue objects (Michael Gorsuch)
* Add #empty? to Queue objects
* Remove deprecated .json extensions from API URIs
* Use new singular route for metric GETs
* README improvements
* Add #clear as alias to Queue's #flush
* Switch to multi_json for better cross-platform json handling
* Set up basic integration testing suite
* Improve testing rake tasks

### Version 0.2.3
* Fix broken user-agent string in 1.8.7 (Sean Porter)
* Update outdated spec

### Version 0.2.2
* Fix abstract persistence instantiation in Ruby 1.8/REE

### Version 0.2.1
* Add better handling for start_time and end_time params when fetching measurements

### Version 0.2.0
* Fix bug with stale excon connections not reconnecting
* Add custom User-Agent
* Items added to Queue objects have their measure_time set automatically
* Metric 'type' key can be string or symbol (Neil Mock)

### Version 0.1.0
* Initial release


lib/librato/metrics/client.rb
module Librato
  module Metrics

    class Client
      extend Forwardable

      def_delegator :annotator, :add, :annotate

      attr_accessor :email, :api_key, :proxy

      # Provide agent identifier for the developer program. See:
      # http://support.metrics.librato.com/knowledgebase/articles/53548-developer-program
      #
      # @example Have the gem build your identifier string
      #   Librato::Metrics.agent_identifier 'flintstone', '0.5', 'fred'
      #
      # @example Provide your own identifier string
      #   Librato::Metrics.agent_identifier 'flintstone/0.5 (dev_id:fred)'
      #
      # @example Remove identifier string
      #   Librato::Metrics.agent_identifier ''
      def agent_identifier(*args)
        if args.length == 1
          @agent_identifier = args.first
        elsif args.length == 3
          @agent_identifier = "#{args[0]}/#{args[1]} (dev_id:#{args[2]})"
        elsif ![0,1,3].include?(args.length)
          raise ArgumentError, 'invalid arguments, see method documentation'
        end
        @agent_identifier ||= ''
      end

      def annotator
        @annotator ||= Annotator.new(client: self)
      end

      # API endpoint to use for queries and direct
      # persistence.
      #
      # @return [String] api_endpoint
      def api_endpoint
        @api_endpoint ||= 'https://metrics-api.librato.com'
      end

      # Set API endpoint for use with queries and direct
      # persistence. Generally you should not need to set this
      # as it will default to the current Librato Metrics
      # endpoint.
      #
      def api_endpoint=(endpoint)
        @api_endpoint = endpoint
      end

      # Authenticate for direct persistence
      #
      # @param [String] email
      # @param [String] api_key
      def authenticate(email, api_key)
        flush_authentication
        self.email, self.api_key = email, api_key
      end

      # Current connection object
      #
      def connection
        # prevent successful creation if no credentials set
        raise CredentialsMissing unless (self.email and self.api_key)
        @connection ||= Connection.new(client: self, api_endpoint: api_endpoint,
                                       adapter: faraday_adapter, proxy: self.proxy)
      end

      # Overrride user agent for this client's connections. If you
      # are trying to specify an agent identifier for developer
      # program, see #agent_identifier.
      #
      def custom_user_agent=(agent)
        @user_agent = agent
        @connection = nil
      end

      def custom_user_agent
        @user_agent
      end

      # Completely delete metrics with the given names. Be
      # careful with this, this is instant and permanent.
      #
      # @example Delete metric 'temperature'
      #   Librato::Metrics.delete_metrics :temperature
      #
      # @example Delete metrics 'foo' and 'bar'
      #   Librato::Metrics.delete_metrics :foo, :bar
      #
      # @example Delete metrics that start with 'foo' except 'foobar'
      #   Librato::Metrics.delete_metrics names: 'foo*', exclude: ['foobar']
      #
      def delete_metrics(*metric_names)
        raise(NoMetricsProvided, 'Metric name missing.') if metric_names.empty?
        if metric_names[0].respond_to?(:keys) # hash form
          params = metric_names[0]
        else
          params = { names: metric_names.map(&:to_s) }
        end
        connection.delete do |request|
          request.url connection.build_url("metrics")
          request.body = SmartJSON.write(params)
        end
        # expects 204, middleware will raise exception otherwise.
        true
      end

      # Return current adapter this client will use.
      # Defaults to Metrics.faraday_adapter if set, otherwise
      # Faraday.default_adapter
      def faraday_adapter
        @faraday_adapter ||= default_faraday_adapter
      end

      # Set faraday adapter this client will use
      def faraday_adapter=(adapter)
        @faraday_adapter = adapter
      end

      # Retrieve measurements for a given composite metric definition.
      # :start_time and :resolution are required options, :end_time is
      # optional.
      #
      # @example Get 5m moving average of 'foo'
      #   measurements = Librato::Metrics.get_composite
      #     'moving_average(mean(series("foo", "*"), {size: "5"}))',
      #     start_time: Time.now.to_i - 60*60, resolution: 300
      #
      # @param [String] definition Composite definition
      # @param [hash] options Query options
      def get_composite(definition, options={})
        unless options[:start_time] && options[:resolution]
          raise "You must provide a :start_time and :resolution"
        end
        query = options.dup
        query[:compose] = definition
        url = connection.build_url("metrics", query)
        response = connection.get(url)
        parsed = SmartJSON.read(response.body)
        # TODO: pagination support
        parsed
      end

      # Retrieve a specific metric by name, optionally including data points
      #
      # @example Get attributes for a metric
      #   metric = Librato::Metrics.get_metric :temperature
      #
      # @example Get a metric and its 20 most recent data points
      #   metric = Librato::Metrics.get_metric :temperature, count: 20
      #   metric['measurements'] # => {...}
      #
      # A full list of query parameters can be found in the API
      # documentation: {http://dev.librato.com/v1/get/metrics/:name}
      #
      # @param [Symbol|String] name Metric name
      # @param [Hash] options Query options
      def get_metric(name, options = {})
        query = options.dup
        if query[:start_time].respond_to?(:year)
          query[:start_time] = query[:start_time].to_i
        end
        if query[:end_time].respond_to?(:year)
          query[:end_time] = query[:end_time].to_i
        end
        unless query.empty?
          query[:resolution] ||= 1
        end
        # expects 200
        url = connection.build_url("metrics/#{name}", query)
        response = connection.get(url)
        parsed = SmartJSON.read(response.body)
        # TODO: pagination support
        parsed
      end

      # Retrieve series of measurements for a given metric
      #
      # @example Get series for metric
      #   series = Librato::Metrics.get_series :requests, resolution: 1, duration: 3600
      #
      # @example Get series for metric grouped by tag
      #   query = { duration: 3600, resolution: 1, group_by: "environment", group_by_function: "sum" }
      #   series = Librato::Metrics.get_series :requests, query
      #
      # @example Get series for metric grouped by tag and negated by tag filter
      #   query = { duration: 3600, resolution: 1, group_by: "environment", group_by_function: "sum", tags_search: "environment=!staging" }
      #   series = Librato::Metrics.get_series :requests, query
      #
      # @param [Symbol|String] metric_name Metric name
      # @param [Hash] options Query options
      def get_series(metric_name, options={})
        raise ArgumentError, ":resolution and :duration or :start_time must be set" if options.empty?
        query = options.dup
        if query[:start_time].respond_to?(:year)
          query[:start_time] = query[:start_time].to_i
        end
        if query[:end_time].respond_to?(:year)
          query[:end_time] = query[:end_time].to_i
        end
        query[:resolution] ||= 1
        unless query[:start_time] || query[:end_time]
          query[:duration] ||= 3600
        end
        url = connection.build_url("measurements/#{metric_name}", query)
        response = connection.get(url)
        parsed = SmartJSON.read(response.body)
        parsed["series"]
      end

      # Retrieve data points for a specific metric
      #
      # @example Get 20 most recent data points for metric
      #   data = Librato::Metrics.get_measurements :temperature, count: 20
      #
      # @example Get 20 most recent data points for a specific source
      #   data = Librato::Metrics.get_measurements :temperature, count: 20,
      #                                            source: 'app1'
      #
      # @example Get the 20 most recent 15 minute data point rollups
      #   data = Librato::Metrics.get_measurements :temperature, count: 20,
      #                                            resolution: 900
      #
      # @example Get data points for the last hour
      #   data = Librato::Metrics.get_measurements start_time: Time.now-3600
      #
      # @example Get 15 min data points from two hours to an hour ago
      #   data = Librato::Metrics.get_measurements start_time: Time.now-7200,
      #                                            end_time: Time.now-3600,
      #                                            resolution: 900
      #
      # A full list of query parameters can be found in the API
      # documentation: {http://dev.librato.com/v1/get/metrics/:name}
      #
      # @param [Symbol|String] metric_name Metric name
      # @param [Hash] options Query options
      def get_measurements(metric_name, options = {})
        raise ArgumentError, "you must provide at least a :start_time or :count" if options.empty?
        get_metric(metric_name, options)["measurements"]
      end

      # Purge current credentials and connection.
      #
      def flush_authentication
        self.email = nil
        self.api_key = nil
        @connection = nil
      end

      # List currently existing metrics
      #
      # @example List all metrics
      #   Librato::Metrics.metrics
      #
      # @example List metrics with 'foo' in the name
      #   Librato::Metrics.metrics name: 'foo'
      #
      # @param [Hash] options
      def metrics(options={})
        query = {}
        query[:name] = options[:name] if options[:name]
        offset = 0
        path = "metrics"
        Collection.paginated_metrics(connection, path, query)
      end

      # Create a new queue which uses this client.
      #
      # @return [Queue]
      def new_queue(options={})
        options[:client] = self
        Queue.new(options)
      end

      # Persistence type to use when saving metrics.
      # Default is :direct.
      #
      # @return [Symbol]
      def persistence
        @persistence ||= :direct
      end

      # Set persistence type to use when saving metrics.
      #
      # @param [Symbol] persist_method
      def persistence=(persist_method)
        @persistence = persist_method
      end

      # Current persister object.
      def persister
        @queue ? @queue.persister : nil
      end

      # Submit all queued metrics.
      #
      def submit(args)
        @queue ||= Queue.new(client: self,
                             skip_measurement_times: true,
                             clear_failures: true)
        @queue.add args
        @queue.submit
      end

      # Update a single metric with new attributes.
      #
      # @example Update metric 'temperature'
      #   Librato::Metrics.update_metric :temperature, period: 15, attributes: { color: 'F00' }
      #
      # @example Update metric 'humidity', creating it if it doesn't exist
      #   Librato::Metrics.update_metric 'humidity', type: :gauge, period: 60, display_name: 'Humidity'
      #
      def update_metric(metric, options = {})
        url = "metrics/#{metric}"
        connection.put do |request|
          request.url connection.build_url(url)
          request.body = SmartJSON.write(options)
        end
      end

      # Update multiple metrics.
      #
      # @example Update multiple metrics by name
      #   Librato::Metrics.update_metrics names: ["foo", "bar"], period: 60
      #
      # @example Update all metrics that start with 'foo' that aren't 'foobar'
      #   Librato::Metrics.update_metrics names: 'foo*', exclude: ['foobar'], display_min: 0
      #
      def update_metrics(metrics)
        url = "metrics" # update multiple metrics
        connection.put do |request|
          request.url connection.build_url(url)
          request.body = SmartJSON.write(metrics)
        end
      end

      # List sources, optionally limited by a name. See http://dev.librato.com/v1/sources
      # and http://dev.librato.com/v1/get/sources
      #
      # @example Get sources matching "production"
      #   Librato::Metrics.sources name: "production"
      #
      # @param [Hash] filter
      def sources(filter = {})
        query = {}
        query[:name] = filter[:name] if filter.has_key?(:name)
        path = "sources"
        Collection.paginated_collection("sources", connection, path, query)
      end

      # Retrieve a single source by name. See http://dev.librato.com/v1/get/sources/:name
      #
      # @example Get the source for a particular EC2 instance from Cloudwatch
      #   Librato::Metrics.get_source "us-east-1b.i-f1bc8c9c"
      #
      # @param String name
      def get_source(name)
        url = connection.build_url("sources/#{name}")
        response = connection.get(url)
        parsed = SmartJSON.read(response.body)
      end

      # Update a source by name. See http://dev.librato.com/v1/get/sources/:name
      #
      # @example Update the source display name for a particular EC2 instance from Cloudwatch
      #   Librato::Metrics.update_source "us-east-1b.i-f1bc8c9c", display_name: "Production Web 1"
      #
      # @param String name
      # @param Hash options
      def update_source(name, options = {})
        url = "sources/#{name}"
        connection.put do |request|
          request.url connection.build_url(url)
          request.body = SmartJSON.write(options)
        end
      end

      # Create a snapshot of an instrument
      #
      # @example Take a snapshot of the instrument at https://metrics-api.librato.com/v1/instruments/42 using a
      #   duration of 3 hours and ending at now.
      #   Librato::Metrics.snapshot(subject: {instrument: {href: "https://metrics-api.librato.com/v1/instruments/42"}},
      #                             duration: 3.hours, end_time: Time.now)
      #
      # @param Hash options Params for the snapshot
      # @options options [Hash] :subject An object representing the subject of the snapshot. For now, only instruments are supported
      # @options options [Numeric] :duration Time interval over which to take the snapshot, defaults to 1 hour
      # @options options [Numeric, Time] :end_time Snapshot the time period of the duration, ending with end_time. Default is "now".
      def create_snapshot(options = {})
        url = "snapshots"
        response = connection.post do |request|
          request.url connection.build_url(url)
          request.body = SmartJSON.write(options)
        end
        parsed = SmartJSON.read(response.body)
      end

      # Retrive a snapshot, to check its progress or find its image_href
      #
      # @example Get a snapshot identified by 42
      #   Librato::Metrics.get_snapshot 42
      #
      # @param [Integer|String] id
      def get_snapshot(id)
        url = "snapshots/#{id}"
        response = connection.get(url)
        parsed = SmartJSON.read(response.body)
      end

    private

      def default_faraday_adapter
        if Metrics.client == self
          Faraday.default_adapter
        else
          Metrics.faraday_adapter
        end
      end

      def flush_persistence
        @persistence = nil
      end

    end

  end
end
